src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Rpc
+{
+    /// <summary>
+    /// Implements the server side of the RPC channel used to communicate with the build host.
+    /// </summary>
+    /// <remarks>
+    /// The RPC system implemented here is pretty close to something like JSON-RPC; however since we need the Build Host to be usable in Source Build
+    /// scenarios, we are limited to using only what is either in .NET or can be easily made buildable in Source Build. Thus existing solutions like StreamJsonRpc 
+    /// are out. If at some point there is a standard RPC mechanism exposed in .NET, we should just delete this and use that instead.
+    /// </remarks>
+    internal sealed class RpcServer
+    {
+        private readonly TextWriter _sendingStream;
+        private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+        private readonly TextReader _receivingStream;
+
+        private readonly ConcurrentDictionary<int, object> _rpcTargets = new();
+        private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
+
+        public RpcServer(Stream sendingStream, Stream receivingStream)
+        {
+            _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+            _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
+        }
+
+        public int AddTarget(object rpcTarget)
+        {
+            // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+            // billions of long lived projects, but...
+            while (true)
+            {
+                var nextIndex = Interlocked.Increment(ref _nextRpcTargetIndex);
+                if (_rpcTargets.TryAdd(nextIndex, rpcTarget))
+                    return nextIndex;
+            }
+        }
+
+        public async Task RunAsync()
+        {
+            var runningTasks = new ConcurrentSet<Task>();
+
+            string? line;
+            while ((line = await _receivingStream.ReadLineAsync().ConfigureAwait(false)) != null)
+            {
+                var request = JsonConvert.DeserializeObject<Request>(line);
+
+                Contract.ThrowIfNull(request);
+
+                var runningTask = Task.Run(async () =>
+                    {
```

Comment:
could you extract this body into a helper.

Reply:
Extracted.